### PR TITLE
Add ResponseHelperCallToJsonResponseRector

### PR DESCRIPTION
**Pull request by Stoffo**

### CODE REVIEW

File: src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Name\FullyQualified;
+use Rector\Rector\AbstractRector;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://laravel.com/docs/7.x/database-testing#creating-models
+ * @changelog https://laravel.com/docs/8.x/database-testing#instantiating-models
```
Comment (GeniJaho) [inline src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:17]: Thanks for the PR :)

These two links are not related to the rule 😅

File: src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @see \RectorLaravel\Tests\Rector\FuncCall\FactoryFuncCallToStaticCallRector\FactoryFuncCallToStaticCallRectorTest
```
Comment (GeniJaho) [inline src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:19]: This `@see` annotation needs to point to the correct test class.

File: src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * @param  MethodCall  $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $this->isName($node->name, 'json')) {
+            return null;
+        }
+
+        if (! $this->isName($node->var, 'response')) {
```
Comment (GeniJaho) [inline src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:56]: This check is not sufficient, since it could work with cases like `$response->json()` when it's not an instance of `JsonResponse`. We should instead make this check for the `ObjectType`.

File: tests/Rector/FuncCall/RemoveRedundantWithCallsRector/Fixture/fixture.php.inc [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 namespace RectorLaravel\Tests\Rector\FuncCall\RemoveRedundantWithCallsRector\Fixture;
 
-use RectorLaravel\Tests\Rector\FuncCall\RemoveRedundantWithCallsRector\Source\HelperObject;
+use RectorLaravel\Tests\Rector\MethodCall\ResponseHelperCallToJsonResponseRector\Source\HelperObject;
```
Comment (GeniJaho) [inline tests/Rector/FuncCall/RemoveRedundantWithCallsRector/Fixture/fixture.php.inc:5]: This class is changed by mistake.

File: tests/Rector/FuncCall/RemoveRedundantWithCallsRector/Fixture/skips_calls_with_callables_as_a_second_arg.php.inc [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 namespace RectorLaravel\Tests\Rector\FuncCall\RemoveRedundantWithCallsRector\Fixture;
 
-use RectorLaravel\Tests\Rector\FuncCall\RemoveRedundantWithCallsRector\Source\HelperObject;
+use RectorLaravel\Tests\Rector\MethodCall\ResponseHelperCallToJsonResponseRector\Source\HelperObject;
```
Comment (GeniJaho) [inline tests/Rector/FuncCall/RemoveRedundantWithCallsRector/Fixture/skips_calls_with_callables_as_a_second_arg.php.inc:5]: This class is changed by mistake.

**Review by GeniJaho — CHANGES_REQUESTED** [review]

File: src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php (hunk 1 above)
Comment (Stoffo) [reply src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php]: Sorry for that, typical copy + paste mistake 😅

File: src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php (hunk 3 above)
Comment (Stoffo) [reply src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:54]: Good point! Didn't think of that. Do you know how to solve that issue? `$this->isObjectType($node, new ObjectType('Illuminate\Routing\ResponseFactory'))` is `false` in that case...

Comment (GeniJaho) [reply src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:54]: The `response()` function returns an instance of `Illuminate\Contracts\Routing\ResponseFactory`, but it may be overkill to check the object type 😅 . It might suffice to just check that it's a FuncCall, with no arguments.

```php
if (! $node->var instanceof FuncCall) {
    return null;
}

if (count($node->var->args) > 0) {
    return null;
}
```

I'm curious, when/why would you use this rule?

Comment (GeniJaho) [reply src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:54]: If you're going the ObjectType way, you'd have to use the `$node->var` in the check, like this `$this->isObjectType($node->var, new ObjectType('Illuminate\Routing\ResponseFactory'))`

File: tests/Rector/MethodCall/ResponseHelperCallToJsonResponseRector/config/configured_rule.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+use Rector\Config\RectorConfig;
+use RectorLaravel\Rector\MethodCall\ResponseHelperCallToJsonResponseRector;
+
+return static function (RectorConfig $rectorConfig): void {
+    $rectorConfig->import(__DIR__ . '/../../../../../config/config.php');
+
+    $rectorConfig->rule(ResponseHelperCallToJsonResponseRector::class);
+};
+
+function response(): ResponseFactory
```
Comment (GeniJaho) [inline tests/Rector/MethodCall/ResponseHelperCallToJsonResponseRector/config/configured_rule.php:15]: Remove this declaration here please. It might conflict with existing `response()` function in the apps, since it has no checks with `function_exists()`. We don't need it for tests to pass (when we're not using the ObjectType check).

**Comment by peterfox** [comment]
It might be better to scale this up to be a rule to handle all response helper scenarios.

Or to make a new configurable rule for the main rector repo. `FuncCallToMethodCallRector` kind of does the same thing but it instead injects a property into a class via the constructor. I'd say a `MethodCallToNew` might make sense.

File: tests/Rector/MethodCall/ResponseHelperCallToJsonResponseRector/config/configured_rule.php (hunk 6 above)
Comment (Stoffo) [reply tests/Rector/MethodCall/ResponseHelperCallToJsonResponseRector/config/configured_rule.php]: Yes I was aware that this was kind of hacky... Was just my first approach to test it. It's removed now

File: src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php (hunk 3 above)
Comment (Stoffo) [reply src/Rector/MethodCall/ResponseHelperCallToJsonResponseRector.php:54]: @GeniJaho Thanks for your input, this fixed the issues!

 The reason I wrote this Rector is that I like this style better than using all the helper functions laravel provides to keep the code more "OOP Style" if you will. It's easier to follow what really happens at first glance.

**Comment by GeniJaho** [comment]
> It might be better to scale this up to be a rule to handle all response helper scenarios.

I don't think it's possible to add them all under one configurable rule, I suspect other `response()` methods require more than just passing parameters to the `new` constructor. We could add them in other rules, and make a set out of it 👀 

> Or to make a new configurable rule for the main rector repo.

A new `MethodCallToNew` in the core repo would probably do the trick, but I'm not sure it can be done in a way that handles different use cases. We have a simple one here, just copy the method params into the class constructor, but how useful/common would that Rector be for other use cases? However, this makes me think we need some core helper rules for common transforms like these...

@peterfox so how do you want to handle this PR? 😅

**Comment by peterfox** [comment]
@GeniJaho I think for now it's not a bad rule. Let's get it in and then if we find it doesn't scale we'll deprecate this for something else.

**Review by GeniJaho — APPROVED** [review]
